doc/fileformat/serdes.rst

```diff
@@ -0,0 +1,262 @@
+.. _formatserdes:
+
+Checkpointing
+=============
+
+While not a fileformat on its own, checkpoints allow users to dump a snapshot of
+the simulation state for a later time. This is **not** intended for data
+extraction (see :ref:`probesample` for this functionality), but instead a reset
+the simulation state to this point in time. The saved data is not a complete
+image of the state, but strives to be a minimal subset from which to restore
+(see :ref:`below <impl guide>`).
+
+Checkpoints are currently _not_ portable; i.e. the following constraints apply:
+
+1. The recipe class must be the same across ``serialize`` / ``deserialize``
+   Unfortunately, we cannot enforce this at runtime, since we lack the
+   appropriate introspection capabilities in C++/Python.
+2. Likewise, ``context`` and ``domain_decomposition`` must be identical. Again
+   we cannot validate this, since we'd either stop migration between ranks or
+   would have to instantiate and sort every local domain decomposition across
+   all ranks.
+3. The hardware configuration must be identical: SIMD, GPU, and Scalar CPU code
+   have different binary layouts which makes serializing across hardware settings
+   invalid.
+
+Thus, the recommended approach is to treat a simulation 'script' (``.cxx`` /
+``.py``), the parallel-runtime parameters (think ``mpirun``) and the associated
+snapshots as a single unit.
+
+.. note:: While this functionality can quite obviously be abused to mutate certain
+          variable out-of-band, we strongly advise against it.
+
+Usage
+-----
+
+Checkpoints are independent of the storage engine used, it allows for multiple
+backends. The general usage is show in the examples below; we assume familiarity
+with the C++ and Python recipe interfaces:
+
+  .. code:: c++
+
+    // Storage engine, pluggable. See below.
+    auto writer = io{};
+    // Serializer object using the engine.
+    auto serializer = arb::serializer{writer};
+
+    // Construct a simulation.
+    auto model = recipe{};
+    auto ctx = arb::make_context();
+    auto ddc = arb::partition_load_balance(model, ctx);
+    auto simulation = arb::simulation{model, ddc, ctx};
+
+    // Run forwards, snapshot state.
+    simulation.run(T, dt);
+    serialize(serializer, simulation);
+    // Then run some more, ...
+    simulation.run(2*T, dt);
+
+    // ... rewind to time T...
+    deserialize(serializer, simulation);
+    // ... and run the same segment again.
+    simulation.run(2*T, dt);
+
+    // At this point the results obtained in the two time segments
+    // [T, 2T) should be identical.
+
+Note that we left the definition of ``io`` open, as ``serializer`` uses it
+through a well-defined interface (see next section). Thus, one can simply add
+new implementations. Arbor currently ships with ``arborio::json_serdes`` that
+produces JSON output, more might be added over time.
+
+In Python, currently only support for (de)serialization from/to JSON strings is
+offered.
+
+  .. code:: python
```

Comment:
I would move the Python demo up-top of the usage section, from which you can naturally proceed to explain C++ offers the option of custom storage engines.

Reply:
Done